osf/models/user.py

```diff
@@ -777,6 +779,32 @@ def merge_user(self, user):
 
         user.save()
 
+    def _merge_users_preprints(self, user):
+        from osf.models.preprint import Preprint, PreprintContributor
+        users_preprint_groups = user.groups.filter(groupobjectpermission__content_type__model=Preprint._meta.verbose_name)
+
+        # If both users are already contribs we will violate unique constraint so take the highest permission group and
+        # just remove old contrib from group and contrib table
+        for preprint_contributor in PreprintContributor.objects.filter(preprint__in=self.preprints.all()):
+            preprint_groups = set(preprint_contributor.preprint.group_objects)
+            shared_groups = preprint_groups.intersection(set(users_preprint_groups))
+            if shared_groups:
+                highest_group = max(shared_groups)
+                self.groups.add(highest_group)
+
+            if preprint_contributor.user == user:
+                preprint_contributor.delete()
```

Comment:
I don't think visibility would necessarily win out in this case. Transferring the highest of two permissions make sense because you are otherwise losing a level of control you previously had, but visibility/invisibility is a matter of preference.

Reply:
Just matching previous behavior on nodes.  When we merge users it wouldn't make sense for nodes to handle visibility one way and preprints another. If more broadly we disagree with how merging behavior should work we should ask product.